### PR TITLE
pkg: add support for git reference repository

### PR DESCRIPTION
**Pull request by kaspar030**
Git allows to specify a reference repository to be specified at clone time, causing all objects available there to be taken from there.

Together with a bare git repository somewhere, that enables transparent git object caching.

GIT_REFERENCE is supposed to have the whole parameter, e.g., ```--reference /home/git-cache.git```

This PR exports the variable, passes it on to buildtest and changes the pkg makefiles accordingly.
It should change nothing unless that variable is set.

Murdock will soon cache all git repos locally.

### CODE REVIEW

File: Makefile.buildtests [hunk 1]
```diff
@@ -51,6 +51,7 @@ buildtest:
 	        BOARD=$${BOARD} \
 	        CCACHE=$${_CCACHE} \
 	        $${CCACHE_DIR:+CCACHE_DIR=$${CCACHE_DIR}} \
+	        $${GIT_REFERENCE:+GIT_REFERENCE="$${GIT_REFERENCE}"} \
```
Comment (cgundogan) [inline Makefile.buildtests]: @OlegHahm: close your eyes (:

Comment (kaspar030) [inline Makefile.buildtests]: hehe

**Comment by cgundogan** [comment]
Hmm, I am not that familiar with git references. At first sight the feature looks nice, but how are cache misses handled? Will git verify first whether a cache is up-to-date or will it just use the cached version until the cache is purge by the user?

BTW, another way to speed up clones would be to use a `--depth=1` or something similar.

**Comment by kaspar030** [comment]
> how are cache misses handled?

git will always contact the remote repository to determine which objects it needs, but then use object data from the reference if available. It's not perfect speed wise, as every clone still needs to ask the network, but it's safe (tm).

**Comment by cgundogan** [comment]
maybe it makes more sense to rename `$(GIT_REFERENCE)` to something like `$(RIOT_GIT_REFERENCE)`. If I have this variable set on my system intentionally that could mess up some things. Having RIOT_ as a prefix makes the usage of that variable more obvious.

**Comment by kaspar030** [comment]
Build succeeded, I've added the caching to murdock, will restart to test.

**Comment by kaspar030** [comment]
Nope. ;) Have to mount that git directory as docker volume...

**Comment by cgundogan** [comment]
I see, you only set it for the docker buildtest. Well okay, untested ACK. The code doesn't change anything if `$(GIT_REFERENCE)` is unset anyhow.

**Comment by cgundogan** [comment]
Hmm, but then again. You modified the pkg's to have this `$(GIT_REFERENCE)`. If I have this on my system defined for other things, this will mess up if I build on my machine (without being in docker)?

**Comment by kaspar030** [comment]
> you only set it for the docker buildtest.

That was unintended, I wanted that for the normal buildtest.

**Comment by kaspar030** [comment]
> If I have this on my system defined for other things, this will mess up if I build on my machine (without being in docker)?

The idea is that if "GIT_REFERENCE" is unset, nothing is changed. If it is set, the variable *should* look like "--reference <path/to/cache/repo>". If that cache repo doesn't contain the main repo, nothing happens.
So unless you have set GIT_REFERENCE to something different, this shouldn't do any harm.

**Comment by cgundogan** [comment]
>So unless you have set GIT_REFERENCE to something different, this shouldn't do any harm.

Yeah that's my point (: IMO `GIT_REFERENCE` is too generic as a name. I can imagine that someone uses that for some other purpose or as a git reference for another project. Naming that `RIOT_GIT_REFERENCE` or `RIOT_PKG_GIT_REFERENCE` would be more specific to RIOT, don't you think?

**Comment by kaspar030** [comment]
> Naming that RIOT_GIT_REFERENCE or RIOT_PKG_GIT_REFERENCE would be more specific to RIOT, don't you think?

Yes, better.

It's not as easy as I thought, anways, as git doesn't like the reference repo when mounted to a different directory using docker. Have to investigate, so I'll mark this as WIP for now.

**Comment by kaspar030** [comment]
closed in favour of #5112.

**Pull request closed without merging**